// yoastseo/src/helpers/getFunctionWords.js
"use strict";

Object.defineProperty(exports, "__esModule", {
	value: true
});

exports.default = function () {
	return {
		en: englishFunctionWords,
		de: germanFunctionWords,
		nl: dutchFunctionWords,
		fr: frenchFunctionWords,
		es: spanishFunctionWords,
		it: italianFunctionWords,
		pt: portugueseFunctionWords,
		ru: russianFunctionWords,
		pl: polishFunctionWords,
		sv: swedishFunctionWords,
		id: indonesianFunctionWords,
		he: hebrewFunctionWords,
		ar: arabicFunctionWords,
		fa: farsiFunctionWords,
		hu: hungarianFunctionWords,
		nb: norwegianFunctionWords,
		tr: turkishFunctionWords
	};
};

var _functionWords = require("../researches/german/functionWords.js");

var _functionWords2 = _interopRequireDefault(_functionWords);

var _functionWords3 = require("../researches/english/functionWords.js");

var _functionWords4 = _interopRequireDefault(_functionWords3);

var _functionWords5 = require("../researches/dutch/functionWords.js");

var _functionWords6 = _interopRequireDefault(_functionWords5);

var _functionWords7 = require("../researches/spanish/functionWords.js");

var _functionWords8 = _interopRequireDefault(_functionWords7);

var _functionWords9 = require("../researches/italian/functionWords.js");

var _functionWords10 = _interopRequireDefault(_functionWords9);

var _functionWords11 = require("../researches/french/functionWords.js");

var _functionWords12 = _interopRequireDefault(_functionWords11);

var _functionWords13 = require("../researches/portuguese/functionWords.js");

var _functionWords14 = _interopRequireDefault(_functionWords13);

var _functionWords15 = require("../researches/russian/functionWords.js");

var _functionWords16 = _interopRequireDefault(_functionWords15);

var _functionWords17 = require("../researches/polish/functionWords.js");

var _functionWords18 = _interopRequireDefault(_functionWords17);

var _functionWords19 = require("../researches/swedish/functionWords.js");

var _functionWords20 = _interopRequireDefault(_functionWords19);

var _functionWords21 = require("../researches/indonesian/functionWords.js");

var _functionWords22 = _interopRequireDefault(_functionWords21);

var _functionWords23 = require("../researches/hebrew/functionWords.js");

var _functionWords24 = _interopRequireDefault(_functionWords23);

var _functionWords25 = require("../researches/arabic/functionWords.js");

var _functionWords26 = _interopRequireDefault(_functionWords25);

var _functionWords27 = require("../researches/farsi/functionWords.js");

var _functionWords28 = _interopRequireDefault(_functionWords27);

var _functionWords29 = require("../researches/hungarian/functionWords.js");

var _functionWords30 = _interopRequireDefault(_functionWords29);

var _functionWords31 = require("../researches/norwegian/functionWords.js");

var _functionWords32 = _interopRequireDefault(_functionWords31);

var _functionWords33 = require("../researches/turkish/functionWords.js");

var _functionWords34 = _interopRequireDefault(_functionWords33);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const germanFunctionWords = (0, _functionWords2.default)(); /*
                                                             * The script collects all the lists of function words per language and returns this collection to a Researcher or a
                                                             * stringProcessing script
                                                             */

const englishFunctionWords = (0, _functionWords4.default)();

const dutchFunctionWords = (0, _functionWords6.default)();

const spanishFunctionWords = (0, _functionWords8.default)();

const italianFunctionWords = (0, _functionWords10.default)();

const frenchFunctionWords = (0, _functionWords12.default)();

const portugueseFunctionWords = (0, _functionWords14.default)();

const russianFunctionWords = (0, _functionWords16.default)();

const polishFunctionWords = (0, _functionWords18.default)();

const swedishFunctionWords = (0, _functionWords20.default)();

const indonesianFunctionWords = (0, _functionWords22.default)();

const hebrewFunctionWords = (0, _functionWords24.default)();

const arabicFunctionWords = (0, _functionWords26.default)();

const farsiFunctionWords = (0, _functionWords28.default)();

const hungarianFunctionWords = (0, _functionWords30.default)();

const norwegianFunctionWords = (0, _functionWords32.default)();

const turkishFunctionWords = (0, _functionWords34.default)();

/**
 * Returns the function words for all languages.
 *
 * @returns {Object} Function words for all languages.
 */
